### PR TITLE
Post-mortem: Deployment on January 17th, 2018

### PR DESCRIPTION
**Pull request by bgeuken**
Today we had some issues after the deployment. This blog explains
what happened.

Co-written with @DavidKang

### CODE REVIEW

**Comment by bgeuken** [comment]
@openSUSE/open-build-service please review

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+</p>
+
+<h4>Diving into the problem</h4>
+
+<p>
+  Between 11:00am and 11:15am we had our regular standup meeting and we informed the rest of the team about the RabbitMQ issue.
+</p>
+
+<p>
+  Right after the standup we meet with the team and discussed the issue. As part of this we briefly went through what we
```
Comment (ChrisBr) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: s/meet/met

Comment (ChrisBr) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: Not sure but doesn't make sense to write in present and past here? meet vs discussed

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+</p>
+
+<p>
+  We split up for a short moment to investigate both scenarios in parallel. One group testing the RabbitMQ setup locally to verify
+  whether the update of the bunny gem caused the exceptions.
+  The other group got in touch with one of OBS admins and the maintainer of the RabbitMQ server.
+</p>
+
+<p>
+  As a result we discovered that the problem was not caused by the servers, but it was caused by a bug in wither the bunny gem or
```
Comment (ChrisBr) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: s/wither/wether

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+<p>
+  Once we realized that there was a good chance that bunny was causing the failure, we checked the bunny changelog and issues.
+  However the changelog was not up to date and thus we missed that there was already a patch and a new released version that solved
+  the issue.
+  So we should have checked the closed issues and recently merged commits as well.
+</p>
+
+<p>
+  Obviously we did not test the bunny update with a similar setup to find out the problem earlier. We probably
```
Comment (ChrisBr) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: Maybe we should investigate a way to get faster packages to deploy?

File: _posts/deployments/2018-01-17-post-mortem-5.html (hunk 2 above)
Comment (alexandergraul) [reply _posts/deployments/2018-01-17-post-mortem-5.html]: s/wither/whether

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+<p>
+  It took roughly 30 minutes until the test suite passed and we could merge the PR. Afterwards we've updated our gem packages and
+  had to wait for the obs-server packages to finish. This usualy takes about an hour.
+  Unfortunately the SLE12 SP3 package build, that we need for deploying new packages on our OBS instance, failed. So we had to
+  trigger a rebuild and wait another hour.
+</p>
+
+<p>
+  While waiting for the packages to finish, a user notified us on <a href="irc://irc.opensuse.org/opensuse-buildservice">IRC</a>,
+  at 1:33pm, that he get's a duplicated email every minute.
```
Comment (alexandergraul) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: s/get's/gets

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+</p>
+
+<p>
+  Apparently we forgot to restart the delayed jobs service after disabling the RabbitMQ setup. Because of that delayed jobs was
+  still trying to send notifications to the RabbitMQ server and failed while trying to.
+  Once we restarted the delayed jobs service the issue was gone. That was at 1:36pm.
+</p>
+
+<p>
+  At 4:10pm we finally had a built package and could deploy to solve the problem and re-enable the RabbitMQ configuration.
```
Comment (alexandergraul) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: "had a built package" -> "had built a package"

File: _posts/deployments/2018-01-17-post-mortem-5.html (hunk 1 above)
Comment (alexandergraul) [reply _posts/deployments/2018-01-17-post-mortem-5.html]: I think it should be consistent (past tense in this case)

**Comment by evanrolfe** [comment]
> What are we going to do about the delayed job part? :-)

How about implementing one step deployment?

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+</p>
+
+<p>
+  Right after the standup we meet with the team and discussed the issue. As part of this we briefly went through what we
+  found out so far and as a result we decided to disable RabbitMQ support in OBS for now.
+  This way we prevented any further failures.
+</p>
+
+<p>
+  After disabling RabbitMQ we concluded that the failure we saw could be caused by two reasons:
```
Comment (evanrolfe) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: its past tense so should say `"could have been caused"`.

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+<p>
+  We split up for a short moment to investigate both scenarios in parallel. One group testing the RabbitMQ setup locally to verify
+  whether the update of the bunny gem caused the exceptions.
+  The other group got in touch with one of OBS admins and the maintainer of the RabbitMQ server.
+</p>
+
+<p>
+  As a result we discovered that the problem was not caused by the servers, but it was caused by a bug in wither the bunny gem or
+  depending amq-protocol gem.
```
Comment (evanrolfe) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: `or **the** depending amp-protocol gem`

File: _posts/deployments/2018-01-17-post-mortem-5.html [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+  At 12:16pm we decided to downgrade bunny for now and created
+  <a href="https://github.com/openSUSE/open-build-service/pull/4363">a pull request</a>.
+</p>
+
+<h4>Seeing the light</h4>
+
+<p>
+  It took roughly 30 minutes until the test suite passed and we could merge the PR. Afterwards we've updated our gem packages and
```
Comment (evanrolfe) [inline _posts/deployments/2018-01-17-post-mortem-5.html]: `"we updated"`

File: _posts/deployments/2018-01-17-post-mortem-5.html (hunk 4 above)
Comment (evanrolfe) [reply _posts/deployments/2018-01-17-post-mortem-5.html]: Alex is write about the ', but it should also be in past tense so `"he was getting"` or `"he was receiving"`

**Comment by hennevogel** [comment]
I think we should meet again and talk about

* one step deployment
* rollback

File: _posts/deployments/2018-01-17-post-mortem-5.html (hunk 5 above)
Comment (bgeuken) [reply _posts/deployments/2018-01-17-post-mortem-5.html]: To me 'had built a package' sounds like we actively built the package, while it was more like waiting for the package to be built.

**Comment by bgeuken** [comment]
>> What are we going to do about the delayed job part? :-)
>
> How about implementing one step deployment?

We manually changed the config and restarted apache, but forgot delayed jobs. I guess we could wrap all this in a script that take care of this.

**Comment by bgeuken** [comment]
I've applied your reviews and added a section to the improvement tasks regarding the delayed jobs issue we had.

File: _posts/deployments/2018-01-17-post-mortem-5.html (hunk 5 above)
Comment (alexandergraul) [reply _posts/deployments/2018-01-17-post-mortem-5.html]: fair enough, maybe the passive form "had a package built" then? To me "had a built package" sounds a bit weird *(what other packages are there? - unbuilt ones?)*.

**Pull request merged**